### PR TITLE
CDP refactor

### PR DESCRIPTION
**Pull request by br648**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This CDP refactor changes trip request processing from daily to hourly. It also reduces the amount of data that is anonymized and uploaded to S3. The following parameter has been added to allow CDP to be disabled:

CONNECTED_DATA_PLATFORM_ENABLED

Prior to this update being deployed, the contents of the TripHistoryUpload table must be removed.

### CODE REVIEW

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (miles-grant-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java:96]: Won't this cause `tripSummaryId` to get out of sync?

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (miles-grant-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: Does this remove the coordinate anonymization feature previously present?

**Review by miles-grant-ibigroup — COMMENTED** [review]
Thanks for these changes! Looks good in my initial testing, just a few small detail questions!

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java:96]: My take on this, and it could be wrong, is that if one of the trip summaries has an error they all will. So, if the first trip summary has an error, record this and then don't process the itineraries or subsequent trip summaries. In this case`tripSummaryId` is redundant.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: No that is still present. The method `ConnectedDataManager.getPlaceCoordinates` works out if the first and last coords should be randomized. These coords are then passed in to `AnonymizedTripRequest` and ultimately into `AnonymizedTripRequest.getPlaceCoordinates` which I've renamed to `AnonymizedTripRequest.getLegCoordinates`.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (miles-grant-ibigroup) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java:96]: Oh yeah this makes sense thanks!

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: @miles-grant-ibigroup the latest refactor does remove the randomizing of coords. We just don't provide them now.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (miles-grant-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: Why doesn't this overflow? Wouldn't this return `2` for a single walking leg?

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (miles-grant-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: The code will correctly also strip locations from a bike or escooter leg (since they are not transit legs) but the comment should reflect this as well

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (miles-grant-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: All non-transit legs should have their locations stripped! Is there any way to use the `isTransit` leg property?

**Review by miles-grant-ibigroup — CHANGES_REQUESTED** [review]
This looks great! Just a few quick questions, although it may be me just misunderstanding things

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: I've updated this along with the comments, hopefully it is clearer now.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: Updated.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: Yep! Updated.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (miles-grant-ibigroup) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: Ah that makes sense now thank you!

**Review by miles-grant-ibigroup — APPROVED** [review]
Looks great, thanks for the changes!

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedLeg.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedLeg.java:37]: Is `rentedVehicle` still needed?

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedItinerary.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedItinerary.java]: Looking at `AnonymizedTripRequest`, this comment should state that the id is unique for the trip summary/request.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: Suffix these two variable names with `Index`.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: Can we just write `if (!leg.transitLeg...`?

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataManager.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataManager.java]: Because this is the only call to this constructor, consider moving the last two arguments into the constructor.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataManager.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataManager.java]: Revert to "Obtain anonymized trip data", or just say "~Obtain~Anonymize trip data".

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataManager.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataManager.java]: Rename to `isConnectedDataPlatformEnabled` and update JavaDoc accordingly.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/TripHistoryUploadJob.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/TripHistoryUploadJob.java:41]: Previously created `TripHistoryUpload` will not have field `uploadHour`, so the job will fail.
What do you recommend to handle existing entries in the `TripHistoryUpload` collection?

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: Rename variable to `isArriveBy`.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: `i <= legs.size() - 1`

File: src/main/java/org/opentripplanner/middleware/utils/Coordinates.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/utils/Coordinates.java:16]: Instead of allowing nulls, would it be better to not provide/persist an instance of `Coordinates` in both `AnonymizedTripRequest` and `AnonymizedLeg` the trip if both `lat` and `lon` are null?

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Looks pretty solid, but I have two blocking comments, one regarding handling of existing upload entries in Mongo, one regarding an index out of bounds both leading to crashes.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedLeg.java
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedLeg.java:37]: I think so yes. Point 5 from Caleb here: https://docs.google.com/document/d/1gELtMoJNxSnhAYilCrCfEIBzlNj9h2pkAEycqX1mh8A/edit#heading=h.44jfqj3j3g8t

Although, if the mode value has to be checked each time to workout the rented vehicle perhaps not. Thoughts?

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedItinerary.java
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedItinerary.java]: The value that is unique for the trip summary/request is the `tripSummaryId` value above this one. I have updated the comment on that parameter instead.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: I could but SonarLint suggests doing it this way because the Boolean value is a class and not a primitive. Happy to change this and others if you think it is cleaner?

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/TripHistoryUploadJob.java
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/TripHistoryUploadJob.java:41]: My understanding is that this never made it passed soak testing so the existing values can be deleted. If that is not the case quite a lot of work will be required to handle it. Let me find out about the first point before I implement the second.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: Good spot, thanks.

File: src/main/java/org/opentripplanner/middleware/utils/Coordinates.java
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/utils/Coordinates.java:16]: That would be cleaner. Updated.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedLeg.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedLeg.java:37]: Not sure about the requirement to check the vehicle type, so this should be fine for now.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedItinerary.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedItinerary.java]: Sure. But then maybe it's best to rename `itineraryId` to `itineraryIndex` because it is more like a sequence number? (and update that field's JavaDoc accordingly)

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: Insert space after comma.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: Remove spaces surrounding `::`.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: This function name is not clear. Maybe rename to `isAnyFirstOrLastLegTransit`, and also have the function scan for all itineraries instead of just the first one?

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: Extract variable for `itineraries.get(0)`

File: src/test/resources/snapshots/org/opentripplanner/middleware/ConnectedDataPlatform/ConnectedDataPlatformTest/canCreateZipFileForTripSummaryWithError-0.json
Comment (binh-dam-ibigroup) [inline src/test/resources/snapshots/org/opentripplanner/middleware/ConnectedDataPlatform/ConnectedDataPlatformTest/canCreateZipFileForTripSummaryWithError-0.json:1]: I don't see a need to serialize `fromPlace` or `toPlace` if the value is null, what do you think? (same applies to those fields for legs).

**Review by binh-dam-ibigroup — COMMENTED** [review]
See the rest of my comments for now, I think we are getting close. Let me know what you decide regarding the old trip upload entries.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (br648) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: @binh-dam-ibigroup It's that a thumbs up to change or leave?!

File: src/test/resources/snapshots/org/opentripplanner/middleware/ConnectedDataPlatform/ConnectedDataPlatformTest/canCreateZipFileForTripSummaryWithError-0.json
Comment (br648) [reply src/test/resources/snapshots/org/opentripplanner/middleware/ConnectedDataPlatform/ConnectedDataPlatformTest/canCreateZipFileForTripSummaryWithError-0.json:1]: I think for a consistent payload it should remain.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: It's fine to follow SonarLint.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/TripHistoryUploadJob.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/TripHistoryUploadJob.java:41]: Any updates on this?

Comment (br648) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/TripHistoryUploadJob.java:41]: No. I'm waiting on Jon and/or Miles. Both are off at the moment.

Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/TripHistoryUploadJob.java:41]: Cool. I'll reassign this issue to you until we hear back.

Comment (br648) [reply src/main/java/org/opentripplanner/middleware/connecteddataplatform/TripHistoryUploadJob.java:41]: Heard back from Miles. This is only deployed th ALT middleware which isn't used in production. I have therefore added to this PR description to delete the older data before deploying.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: Extract variable for `coords.split`

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java
Comment (binh-dam-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/AnonymizedTripRequest.java]: Insert space before 'minus'.

**Review by binh-dam-ibigroup — APPROVED** [review]
Good to go, let's coordinate when you are ready to merge (merging will autodeploy, so we'll have to add the environment variable and update Mongo before hand).

**Comment by miles-grant-ibigroup** [comment]
Environment variables are updated -- I tried to update Mongo but couldn't get access to the relevant databases.

**Comment by br648** [comment]
Tests are failing, I think because OTP isn't available: `OtpDispatcherResponse{statusCode=***00, responseBody='<html><body><h2>***00 Internal Server Error</h2></body></html>'PARSE_EXCEPTION}`

@binh-dam-ibigroup Any idea how to fix?

**Comment by binh-dam-ibigroup** [comment]
> Tests are failing, I think because OTP isn't available: `OtpDispatcherResponse{statusCode=***00, responseBody='<html><body><h2>***00 Internal Server Error</h2></body></html>'PARSE_EXCEPTION}`
> 
> @binh-dam-ibigroup Any idea how to fix?

@br648 Not sure, but it seems that the first error in the e2e tests is about an unmocked OTP request, see https://github.com/ibi-group/otp-middleware/runs/7285669028?check_suite_focus=true#step:9:301. Could you check how this is happening?

**Comment by miles-grant-ibigroup** [comment]
The OTP mock server was having some trouble. Now that this has been resolved the tests pass. I will hold off on merging until we've wiped the appropriate databases.

**Pull request merged**
